thevoid/connection.cpp

```diff
@@ -83,6 +86,76 @@ struct timespec gettime_now() {
 
 } // unnamed namespace
 
+
+namespace {
+
+template <typename OutputIterator>
+struct headers_grammar
+	: boost::spirit::karma::grammar<OutputIterator, std::vector<ioremap::swarm::headers_entry>()>
+{
+	typedef std::vector<ioremap::swarm::headers_entry> headers_type;
+	typedef headers_type::value_type header_type;
+	typedef header_type::first_type header_key_type;
+	typedef header_type::second_type header_value_type;
+
+	headers_grammar()
+		: headers_grammar::base_type(headers)
+	{
+		// Mapping of escaped characters
+		escaped_char.add
+			('\a', "\\a")
+			('\b', "\\b")
+			('\f', "\\f")
+			('\n', "\\n")
+			('\r', "\\r")
+			('\t', "\\t")
+			('\v', "\\v")
+			('\\', "\\\\")
+			('\'', "\\\'")
+			('"', "\\\"")
+			;
+
+		// Header value is sequence of:
+		// - either escaped character
+		// - or printable character
+		// - or hexadecimal representation of the corresponding character
+		header_value = *(escaped_char | boost::spirit::karma::print | "\\x" << boost::spirit::karma::hex);
+
+		// Header key is just sequence of characters with no escaping
+		header_key = boost::spirit::karma::string;
+
+		// Header is header key and value separated with ': ' and enclosed in double quotes
+		header = '"' << header_key << ": " << header_value << '"';
+
+		// Headers is (optionally empty) list of headers separated with ', ' and enclosed in {}
+		headers = '{' << -(header << *(", " << header)) << '}';
+	}
+
+	boost::spirit::karma::symbols<char, const char*> escaped_char;
+
+	boost::spirit::karma::rule<OutputIterator, header_value_type()> header_value;
+	boost::spirit::karma::rule<OutputIterator, header_key_type()> header_key;
+
+	boost::spirit::karma::rule<OutputIterator, header_type()> header;
+
+	boost::spirit::karma::rule<OutputIterator, headers_type()> headers;
+};
+
+std::string headers_to_string(const ioremap::swarm::http_headers& headers) {
+	typedef std::back_insert_iterator<std::string> sink_type;
+
+	std::string headers_string;
+	sink_type sink(headers_string);
+
+	headers_grammar<sink_type> generator;
```

Comment:
Is it really necessary to initialize new generator for each request? It could probably be better to have one static generator.